File: packages/auth/src/required-mfa-policy.test.ts
import { describe, expect, it } from 'vitest';
import {
  mfaProofFromVerifiedClaims,
  readRequiredMfaPolicy,
  satisfiesRequiredMfaPolicy,
} from './required-mfa-policy';

const now = 1_790_000_000;
const claims = {
  aal: 'aal2',
  session_id: 'session',
  amr: [{ method: 'totp', timestamp: now - 10 }],
};

describe('required MFA policy', () => {
  it('preserves accounts without a required policy', () => {
    expect(satisfiesRequiredMfaPolicy(readRequiredMfaPolicy({}), null)).toBe(
      true
    );
  });

  it.each([
    null,
    undefined,
    true,
    'true',
    [],
    {},
    { required: 'false' },
    { required: true },
    { required: true, verifiedAfter: -1 },
    { required: true, verifiedAfter: '123' },
    { required: true, verifiedAfter: Number.NaN },
  ])('rejects malformed policy %j', (policy) => {
    expect(
      satisfiesRequiredMfaPolicy(
        readRequiredMfaPolicy({ tuturuuu_required_mfa: policy }),
        mfaProofFromVerifiedClaims(claims, now),
        now
      )
    ).toBe(false);
  });

  it('requires a proof after the latest enforcement or recovery boundary', () => {
    const policy = readRequiredMfaPolicy({
      tuturuuu_required_mfa: { required: true, verifiedAfter: now - 5 },
    });
    expect(
      satisfiesRequiredMfaPolicy(
        policy,
        mfaProofFromVerifiedClaims(claims, now),
        now
      )
    ).toBe(false);
    expect(
      satisfiesRequiredMfaPolicy(
        policy,
        { sessionId: 'session', verifiedAt: now },
        now
      )
    ).toBe(true);
    expect(
      satisfiesRequiredMfaPolicy(
        policy,
        { sessionId: 'session', verifiedAt: now + 1 },
        now
      )
    ).toBe(false);
  });

  it('disabling the policy removes the additional assurance requirement', () => {
    expect(
      satisfiesRequiredMfaPolicy(
        readRequiredMfaPolicy({
          tuturuuu_required_mfa: { required: false, verifiedAfter: now },
        }),
        null,
        now
      )
    ).toBe(true);
  });

  it.each([
    { ...claims, aal: 'aal1' },
    { ...claims, session_id: '' },
    { ...claims, amr: undefined },
    { ...claims, amr: [{ method: 'password', timestamp: now }] },
    { ...claims, amr: [{ method: 'totp', timestamp: now + 1 }] },
    { ...claims, amr: [{ method: 'totp', timestamp: '123' }] },
  ])('rejects missing or invalid verified MFA evidence %j', (input) => {
    expect(mfaProofFromVerifiedClaims(input, now)).toBeNull();
  });

  it.each(['totp', 'mfa/phone', 'mfa/webauthn', 'mfa/recovery_code'])(
    'accepts provider-verified second factor %s',
    (method) => {
      expect(
        mfaProofFromVerifiedClaims(
          { ...claims, amr: [{ method, timestamp: now }] },
          now
        )
      ).toEqual({ sessionId: 'session', verifiedAt: now });
    }
  );

  it('does not extend an expired mobile approval during app refresh', () => {
    const policy = { required: true, verifiedAfter: now - 100 };
    const proof = {
      sessionId: 'session',
      verifiedAt: now - 10,
      expiresAt: now,
    };
    expect(satisfiesRequiredMfaPolicy(policy, proof, now - 1)).toBe(true);
    expect(satisfiesRequiredMfaPolicy(policy, proof, now)).toBe(false);
  });

  it('refreshing a token does not refresh the MFA proof', () => {
    expect(mfaProofFromVerifiedClaims({ ...claims, iat: now }, now)).toEqual({
      sessionId: 'session',
      verifiedAt: now - 10,
    });
  });
});

File: apps/mobile/test/features/auth/required_mfa_policy_test.dart
import 'dart:convert';

import 'package:flutter_test/flutter_test.dart';
import 'package:mobile/features/auth/required_mfa_policy.dart';
import 'package:supabase_flutter/supabase_flutter.dart';

Session session({
  required Map<String, dynamic> metadata,
  int? proofTime,
  int? primaryTime,
  int? expiresAt,
}) {
  final claims = {
    'exp': expiresAt ?? DateTime.now().millisecondsSinceEpoch ~/ 1000 + 3600,
    'aal': proofTime == null ? 'aal1' : 'aal2',
    'session_id': 'session',
    'amr': [
      if (proofTime != null) {'method': 'totp', 'timestamp': proofTime},
      if (primaryTime != null) {'method': 'password', 'timestamp': primaryTime},
    ],
  };
  final token =
      'header.${base64Url.encode(utf8.encode(jsonEncode(claims)))}.signature';
  return Session.fromJson({
    'access_token': token,
    'token_type': 'bearer',
    'refresh_token': 'test',
    'user': {
      'id': 'actor',
      'aud': 'authenticated',
      'created_at': '2026-09-23T00:00:00Z',
      'app_metadata': metadata,
      'user_metadata': <String, dynamic>{},
    },
  })!;
}

void main() {
  test('recovery requires fresh primary authentication before new MFA', () {
    final metadata = {
      'tuturuuu_required_mfa': {
        'required': true,
        'verifiedAfter': 100,
        'primaryVerifiedAfter': 100,
      },
    };
    expect(
      requiresFreshPrimaryForMfa(
        session(metadata: metadata, proofTime: 102, primaryTime: 99),
      ),
      isTrue,
    );
    expect(
      requiresAccountMfa(
        session(metadata: metadata, proofTime: 102, primaryTime: 99),
      ),
      isTrue,
    );
    expect(
      requiresAccountMfa(
        session(metadata: metadata, proofTime: 102, primaryTime: 101),
      ),
      isFalse,
    );
  });
  test('in-progress recovery blocks otherwise fresh cached proofs', () {
    expect(
      requiresAccountMfa(
        session(
          metadata: {
            'tuturuuu_required_mfa': {
              'required': true,
              'verifiedAfter': 100,
              'recoveryInProgress': true,
            },
          },
          proofTime: 102,
          primaryTime: 101,
        ),
      ),
      isTrue,
    );
  });
  test('expired required proof gates cached navigation until refresh', () {
    final now = DateTime.now().millisecondsSinceEpoch ~/ 1000;
    expect(
      requiresAccountMfa(
        session(
          metadata: {
            'tuturuuu_required_mfa': {
              'required': true,
              'verifiedAfter': now - 60,
            },
          },
          proofTime: now - 30,
          expiresAt: now - 1,
        ),
      ),
      isTrue,
    );
  });
  test(
    'optional cached session is not blocked by required-policy expiry logic',
    () {
      expect(requiresAccountMfa(session(metadata: {}, expiresAt: 1)), isFalse);
    },
  );

  test(
    'factorless required accounts must enroll before ordinary navigation',
    () {
      expect(
        requiresAccountMfa(
          session(
            metadata: {
              'tuturuuu_required_mfa': {'required': true, 'verifiedAfter': 100},
            },
          ),
        ),
        isTrue,
      );
    },
  );
  test(
    'reset invalidates previous verification without trusting refresh time',
    () {
      final metadata = {
        'tuturuuu_required_mfa': {'required': true, 'verifiedAfter': 100},
      };
      expect(
        requiresAccountMfa(session(metadata: metadata, proofTime: 100)),
        isTrue,
      );
      expect(
        requiresAccountMfa(session(metadata: metadata, proofTime: 101)),
        isFalse,
      );
    },
  );
  test(
    'missing policy preserves existing routing; malformed policy fails closed',
    () {
      expect(requiresAccountMfa(session(metadata: {})), isFalse);
      expect(
        requiresAccountMfa(session(metadata: {'tuturuuu_required_mfa': null})),
        isTrue,
      );
    },
  );
}
